### PR TITLE
[metric package] New way of reducing

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
I was asked on slack to provide some comments to the new metric package. This PR account for some of my thoughts. Some explanation:
* Accuracy: In the main branch for the native torch metrics, accuracy calculation only works in the binary case because the `(tps + tns) / (tps + tns + fps + fns)` formula does not generalize to multilabel. This has been correct.

* Support: In general the accuracy can be calculated as `tps/support` where the support is the number of times a class is present in the target. This has been added to the `stat_scores_multiple_classes` function.

* Reduction: The way of reducing on the main branch does not make sense, for example it does not make sense to sum the accuracies for individual classes since this will probably give a score larger than 1. Currently it follows the format of the loss functions from `torch.nn` but I think it should instead follow the format from sklearn. Reduction now has the following modes (naming should probably be different, directly taken from sklearn):
    1. 'micro' = global metric (what most users wants)
    2. 'macro' = calculate metrics for each label, and find their unweighted mean
    3. 'weighted' = calculate metrics for each label, and find their weighted mean (for dealing with class-inbalance)
    4. 'none' or None = return measure for individual classes

* Testing: I have added a single (parametrized) test, that test the implemented torch metrics directly against the sklearn counterpart on random tensors. This should secure that the metrics calculated correctly in the future.

@Borda and @justusschock take a look and see if you agree :)

### CODE REVIEW

**Comment by justusschock** [comment]
Not completely sure I like this. With my reduction I tried to stick to the torch.nn.Module reduction interface (like in torch.nn.MSELoss). We can maybe extend this function, but I'd rather not have it replaced...

**Comment by Borda** [comment]
@justusschock so you have some suggestion for a change or rather completely drop this change?

**Comment by justusschock** [comment]
@Borda I'd prefer a solution that has an interface like:

```
def reduce(to_reduce: torch.Tensor, denom: Optional[torch.Tensor] = None, reduction: Optional[str] = None)
```

if the reduction value is set to one of the previous values (e.g. `elementwise_mean`, the denom will be ignored, and if it is set to one of the new values (e.g. `micro`) it is required

**Comment by SkafteNicki** [comment]
I think there is a fundamentally difference between the torch.nn.Module way of reducing and the way that we are reducing metrics. In the torch.nn.Module reduction interface, the reduction parameter determines how to reduce over multi dimensional tensor whereas for the metrics package we are discussion how to reduce over classes.

**Comment by justusschock** [comment]
@SkafteNicki For classification tasks you're for sure right, but for segmentation tasks for example you may also need spatial reduction, which is why I'd like to also have this option.

**Comment by SkafteNicki** [comment]
@justusschock then maybe we need to have a global `reduction` parameter for all functionals and a special `class_reduction` argument for functionals that works with discrete output.

**Comment by SkafteNicki** [comment]
Just to be sure is the intention that we should support multi-dim tensors similar to `torch.nn.module` loss functions? By this I mean that if the user inputs a `pred` and `target` (both as dense labels) with shape `[B,N]` and `reduce=none` then the output should be a `[N,]` tensor.

**Comment by justusschock** [comment]
@SkafteNicki the function does not (yet) know about the type of labels. I.e. there is no such thing as dense labels.

If you stick to my segmentation example, in 2D `pred` and `target` would both be of shape `[N, (C,) H, W]` and so should the output if it isn't reduced...

**Comment by SkafteNicki** [comment]
@justusschock what I mean by dense, is that `pred` is tensor of labels and not a tensor with logits/softmax probabilities. In this case `pred.ndim==target.ndim+1`.

**Comment by Borda** [comment]
@SkafteNicki how is it going here? :]

**Comment by SkafteNicki** [comment]
@Borda is completely forgot about this. I think we ended up not completely agreeing how the reduction interface should be exactly. I think that it would be great to keep it consist with how `reduction` is done in `torch.nn` loss functions. However, I do not think this is feasible since loss functions follows the very strict requirement that it can be decomposed into individual terms for each sample (which is the very reason we can do mini-batching). Therefore, it make sense that if `reduction='none'` for a tensor of size `[N, C, H, W]` (if we are doing segmentation) that the loss output is a `[N, C, H, W]` tensor.

Metrics, however, are defined as global measures and not all of them can be decomposed into individual terms for each samples, with accuracy being an exception. Therefore, for the metric interface I think it is much more sensible for the same example of a `[N, C, H, W]` tensor to reduce to a `[C, H, W]` tensor if `reduction='none'`.

Additionally, I think that we should have a `class_reduction` argument, that determines how to reduce over the different classes (similar to sklearn). 

Right now, I feel like closing this PR and then come back to this after my work on the tuner class (as we discussed on slack the other day).

**Comment by Borda** [comment]
sure, so let's close this one and start with the fresh state
hopefully, merged `native_torch_metrics` to `master` :rabbit:
